igor-web/src/main/java/com/netflix/spinnaker/igor/gcb/GoogleCloudBuildCache.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2019 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.igor.gcb;
+
+import com.netflix.spinnaker.igor.polling.LockService;
+import com.netflix.spinnaker.kork.jedis.RedisClientDelegate;
+import lombok.AccessLevel;
+import lombok.RequiredArgsConstructor;
+
+import java.time.Duration;
+import java.util.Map;
+
+/**
+ * Cache to keep track of the status of Google Cloud builds. In general, this cache will be updated as echo
+ * receives PubSub build notifications and sends them to igor.
+ */
+@RequiredArgsConstructor(access=AccessLevel.PRIVATE)
+public class GoogleCloudBuildCache {
+  private final LockService lockService;
+  private final RedisClientDelegate redisClientDelegate;
+  private final String keyPrefix;
+  private final String lockPrefix;
```

Comment:
Redis for dummies question, what is the purpose of the lock prefix?

Reply:
The lock works by setting a key in redis.  The algorithm is roughly:
* Use the redis "set if not exist" to try to set the key
* If the operation fails, it means the key exists and someone else has the lock; if it succeeds, this process go the lock
* Do the requested work then delete the key

The prefix is the prefix for that actual key we're creating.  We create a key of format `googleCloudBuild.account.buildId` so that the lock is unique to each account/build combo (with the prefix just to have some namespacing).

As we create a `GoogleCloudBuildCache` for each account, the `lockPrefix` there is set to `googleCloudBuild.account` and we just need to append the `buildId` each time we want to get the lock.

Reply:
:+1: thanks for the great explanation!